### PR TITLE
Folder output

### PR DESCRIPTION
**Pull request by gamazeps**
Writes the output to a given folder, and the: `packageName.benchName.json` file in the folder provided by -`-output`, in the `example/Main.hs` file this would be hyperion-example.example.json, also preprends the benchmark name by packageName.

Note that `packageName` is obtained by `getProgName` which is the name of the executable that launched the tests

The objective is to allow a user to have multiple benchmarks in a project (in their different packages), with each one in a separate file, all the while avoiding key collisions.

### CODE REVIEW

File: src/Hyperion/Main.hs [hunk 1]
```diff
@@ -130,22 +134,26 @@ doRun bks = do
       throwIO $ DuplicateNames [ n | n:_:_ <- group (sort nms) ]
     foldMap (runBenchmark (sample 100 (fixed 5))) bks
 
-doAnalyze :: Config -> [Benchmark] -> IO ()
-doAnalyze Config{..} bks = do
+doAnalyze :: Config -> Text -> [Benchmark] -> IO ()
+doAnalyze Config{..} benchName bks = do
+    packageName <- getProgName
```
Comment (qnikst) [inline src/Hyperion/Main.hs]: I do now understand this bit, this actually an executable name, not package name, and each package may contain many such executables. I see that roles of the `getProgName` and `benchName` differ, but I'm not sure how to use them properly, maybe add some documentation about it?

**Comment by qnikst** [comment]
Tests failed with:
```
  1) Hyperion.Main.defaultMain Analyzes uniquely named benchmarks
       uncaught exception: IOException of type AlreadyExists (/dev/null: createDirectory: already exists (File exists)) (after 1 test)
       Series [1,2,3] (\_ -> Bracket (\_ -> Group "two" []))
```

File: src/Hyperion/Main.hs (hunk 1 above)
Comment (gamazeps) [inline src/Hyperion/Main.hs]: You are right the name is the one of the executable containing the bench.

This is done so that one can have multiple benches in an executable, and multiple executables in a same project.

If you have a better naming idea I'm all up for it :)

**Comment by gamazeps** [comment]
@qnikst yup, since I changed the name of the keys by prepending the executable name before it the test do not pass anymore, I wanted to check that with someone face to face tomorrow.

Otherwise the behaviour is indeed the one we would expect, the tests just need to be changed (because the expected behaviour they tested has changed).

File: src/Hyperion/Analysis.hs [hunk 2]
```diff
@@ -49,11 +49,12 @@ namesOf = go []
...
 analyze
-  :: Text -- ^ Benchmark name
+  :: Text -- ^ Package name
+  -> Text -- ^ Benchmark name
   -> Sample -- ^ Measurements
   -> Report
-analyze name samp = Report
-    { _reportBenchName = name
+analyze packageName name samp = Report
+    { _reportBenchName = Text.concat [packageName, ":", name]
```
Comment (mboes) [inline src/Hyperion/Analysis.hs]: Why is this necessary?

File: src/Hyperion/Main.hs [hunk 3]
```diff
@@ -78,7 +81,8 @@ options = do
          (Options.strOption
             (Options.long "output" <>
              Options.short 'o' <>
-             Options.metavar "FILENAME"))
+             Options.help "Outputs the benchark to the directory" <>
```
Comment (mboes) [inline src/Hyperion/Main.hs]: Change to "Where to write the benchmarks output. Can be a directory name." Options don't act.

File: src/Hyperion/Main.hs [hunk 4]
```diff
@@ -130,22 +134,26 @@ doRun bks = do
...
+      Just path -> do
+        let filename = executableName <.> (unpack benchName) <.> "json"
+        createDirectoryIfMissing True path -- Creates the directory if needed.
+        IO.openFile (path </> filename) IO.WriteMode
     results <- doRun bks
     let strip
           | configRaw = id
           | otherwise = reportMeasurements .~ Nothing
-        report = results & imapped %@~ analyze & mapped %~ strip
+        report = results & imapped %@~ (analyze (pack executableName)) & mapped %~ strip
```
Comment (mboes) [inline src/Hyperion/Main.hs]: This looks incorrect to me: you're passing the `executableName` as the `packageName`.

File: src/Hyperion/Main.hs (hunk 3 above)
Comment (gamazeps) [inline src/Hyperion/Main.hs]: It actually has to be a directory because we have no way to know whether the user gives us a directory or a file (if they don't already exist).

File: src/Hyperion/Analysis.hs [hunk 5]
```diff
@@ -49,11 +49,13 @@ namesOf = go []
...
-  -> Sample -- ^ Measurements
+  :: Text -- ^ Executable name.
+  -> Text -- ^ Benchmarks name.
+  -> Text -- ^ Benchmark name.
+  -> Sample -- ^ Measurements.
   -> Report
-analyze name samp = Report
-    { _reportBenchName = name
+analyze executableName benchesName name samp = Report
+    { _reportBenchName = Text.concat [executableName, ".", benchesName, ".", name]
```
Comment (mboes) [inline src/Hyperion/Analysis.hs]: I repeat my comment for a previous version of this: why are you prefixing with the executable name? What is "benchmarks name"?

File: src/Hyperion/Analysis.hs (hunk 2 above)
Comment (gamazeps) [inline src/Hyperion/Analysis.hs]: Has been changed.

This is needed to avoid key collisions for the json created by the benchmarks.
Indeed If I have two benches with the same name they would generate the same names, on the other hand the executable name is more likely to be unique,

File: src/Hyperion/Analysis.hs (hunk 5 above)
Comment (gamazeps) [inline src/Hyperion/Analysis.hs]: Benchmark name is the name passed to `defaultMain`.

As you remarked before, a user can call this function multiple times from a same benchmark executable and can also have multiple benchmark executables in their project.

Taking the two allows to avoid collisions.

File: src/Hyperion/Main.hs (hunk 3 above)
Comment (mboes) [inline src/Hyperion/Main.hs]: That's not true, and has been discussed before. I'd rephrase that as "I don't know how to tell whether the user gave us a directory or a file". You can use `doesDirectoryExist` from the `directory` package for this. If the path does not exist, then assume it's a file. If it does exist, and is a directory, then it's a directory. Optional refinement: use `hasTrailingPathSeparator` in `filepath` to create a directory if it doesn't exist, because in that case it's clear that the user does want a directory.

`rsync` and other standard unix tools use a similar convention.

File: src/Hyperion/Analysis.hs (hunk 5 above)
Comment (mboes) [inline src/Hyperion/Analysis.hs]: That's not exactly what I remarked. Let's at least agree on some nomenclature here:

* Cabal package name = the name in the .cabal (passed manually by the user).
* Cabal benchmark name = the name of the benchmark executable (inferred via `getProgName`)
* Benchmark group = name passed to `bgroup`.
* benchmark name (or "case name" to disambiguate) = name passed to `bench`.

A project can have multiple Cabal packages, each one of which can have multiple Cabal benchmark names. Each Cabal benchmark can define multiple benchmark cases, optionally grouped together in some hierarchy internal to that benchmark suite.

A user should only only ever call `defaultMain` _once_ in any Cabal benchmark. That's why it's called `*Main` after all.

**Comment by mboes** [comment]
Key collisions where? In each .json file I don't believe this should
happen. In ElasticSearch? Yes quite possibly.

But just how unique should the keys really be? Because package names aren't
necessarily unique across projects. So then we'd need to include the
project name too when loading into ES etc. I think this change is unrelated
to the main objective of the patch, so I think we should roll it back and
discuss it separately.

File: src/Hyperion/Analysis.hs (hunk 5 above)
Comment (gamazeps) [inline src/Hyperion/Analysis.hs]: So when we added the Text argument to `defaultMain` it was for the user to specify the package right ?

We would thus want the following naming: `$packageName.$exectuableName.json` right ?

File: src/Hyperion/Main.hs (hunk 3 above)
Comment (gamazeps) [inline src/Hyperion/Main.hs]: Th reason why we wanted to have a directory in the first place is to deal with the fact that the same arguments are passed to all benchmarks executable (defined in `.cabal`), thus providing a file would have them overwrite each other, as each benchmark executable would try to write to the same file.

The `hasTrailingPathSeparator` indeed sounds like the proper solution because it would allow a user to output its content to a folder without having to create it beforehand.

Having a predefined naming convention also seemed like a win because it makes parsing easier and allows an simple output name for raw output (packageName.executableName.raw.json)

Does the following behaviour suit you:
- If output is a directory that exists, output to the directory with the naming convention
- else
  - if `hasTrailingPathSeparator`, create the directory and write to it
  - otherwise output to teh file and create it if needed.

**Comment by gamazeps** [comment]
The key collision I wanted to avoid was indeed in ES.

We discussed this solution, and the conclusion was that we wanted to avoid key collision in ES as much as possible.

This made sense in this PR as the objective was to avoid collisions in benchmark outputs, as the previous behaviour created file collisions it seems logical to also remove key collision in ES.

**Comment by gamazeps** [comment]
@mboes PTAL to the comments

File: src/Hyperion/Analysis.hs (hunk 5 above)
Comment (mboes) [inline src/Hyperion/Analysis.hs]: Yes, that's right.

File: src/Hyperion/Main.hs (hunk 3 above)
Comment (mboes) [inline src/Hyperion/Main.hs]: > thus providing a file would have them overwrite each other, as each benchmark executable would try to write to the same file.

That's true, but it's also not our problem. We shouldn't be assuming the user will use Stack, or indeed even `cabal-install`. And the common case really is that the user has one project, containing one package, with one benchmark suite. Benchmarks are just binaries like any other, so we should make the common case easy, the general case possible.

I don't understand what the diff is between your proposed behaviour and the one I wrote above. Aren't they precisely the same?

**Comment by mboes** [comment]
I remember now that we discussed it. But key collisions in ES (which could in any case be avoided by other means) have nothing to do with file collisions locally within a project. If you remove that particular change, we can proceed with merging the rest of the changes faster, while we hash out what should be done for ES. PR's are cheap, so nothing wrong with two concurrent PR's.

File: src/Hyperion/Main.hs (hunk 3 above)
Comment (gamazeps) [inline src/Hyperion/Main.hs]: It is kinda our problem because if the user does use `stack` this will not work, and *we* use stack.

The behaviour is what i understood of what you said, just wanted to meke sure that we understood each other.

**Comment by gamazeps** [comment]
Will remove

**Comment by gamazeps** [comment]
@mboes all comments are addressed PTAL

File: src/Hyperion/Main.hs [hunk 6]
```diff
@@ -141,11 +146,28 @@ doRun bks = do
...
 
-doAnalyze :: Config -> [Benchmark] -> IO ()
-doAnalyze Config{..} bks = do
+doAnalyze
+  :: Config -- ^ Hyperion config.
+  -> Text -- ^ Package name.
+  -> [Benchmark] -- ^ Benchmarks to be run.
+  -> IO ()
+doAnalyze Config{..} packageName bks = do
+    executableName <- getProgName -- Name of the execuatble that launched the benches.
```
Comment (mboes) [inline src/Hyperion/Main.hs]: typo

File: src/Hyperion/Main.hs [hunk 7]
```diff
@@ -141,11 +146,28 @@ doRun bks = do
...
     h <- case configOutputPath of
       Nothing -> return IO.stdout
-      Just path -> IO.openFile path IO.WriteMode
+      Just path ->
+        if path == nullOutputPath
+          then IO.openFile nullOutputPath IO.WriteMode
```
Comment (mboes) [inline src/Hyperion/Main.hs]: See style guide for proper indentation of `if`-`then`-`else`.

Comment (mboes) [inline src/Hyperion/Main.hs]: You shouldn't need to special case `nullOutputPath`. Notice that the action is the same is in the `else` branch of the second if below.

File: src/Hyperion/Main.hs [hunk 8]
```diff
@@ -169,7 +195,10 @@ defaultMainWith presetConfig bks = do
         Version -> putStrLn $ "Hyperion " <> showVersion version
         List -> doList bks
         Run -> do _ <- doRun bks; return ()
-        Analyze -> doAnalyze config bks
+        Analyze -> doAnalyze config (pack packageName) bks
 
-defaultMain :: [Benchmark] -> IO ()
+defaultMain
+  :: String -- ^ Packahe name, user provided.
```
Comment (mboes) [inline src/Hyperion/Main.hs]: typo

File: stack.yaml [hunk 9]
```diff
@@ -1,4 +1,5 @@
 resolver: lts-7.14
+
```
Comment (mboes) [inline stack.yaml]: unrelated whitespace change.

File: src/Hyperion/Main.hs (hunk 8 above)
Comment (qnikst) [inline src/Hyperion/Main.hs]: Also please keep :: on the same row as function name, this improve experience when you grep for function definition, just search `function ::`

Comment (mboes) [inline src/Hyperion/Main.hs]: Well, but @gamazeps is just following the style guide here. And a multi-line sig is inevitable if you have Haddock annotations.

Comment (qnikst) [inline src/Hyperion/Main.hs]: Should I rise an issue for the guide? 
```
defaultMain ::
    String ^ -- comment
```
Or
```
defaultMain :: String
      -- ^ comment
```
Would not work?

File: stack.yaml (hunk 9 above)
Comment (gamazeps) [inline stack.yaml]: yes, but needed nevertheless

Comment (mboes) [inline stack.yaml]: ?

File: src/Hyperion/Main.hs (hunk 8 above)
Comment (mboes) [inline src/Hyperion/Main.hs]: Not without breaking alignment with arrows. Isn't `^foo` just as easy to search for?

File: stack.yaml (hunk 9 above)
Comment (gamazeps) [inline stack.yaml]: Not having a whitespace there seems wrong, even though it is indeed unrelated, but i'm nitpicking here.

Comment (mboes) [inline stack.yaml]: https://github.com/tweag/guides/blob/master/style/Haskell.md#general-style, last paragraph of that section. Not that it's worth making a point about this one, but do be aware of that.

Comment (gamazeps) [inline stack.yaml]: done

Comment (gamazeps) [inline stack.yaml]: off to sleep, will merge tomorrow morning

Comment (mboes) [inline stack.yaml]: Please send it back to me for review.

**Comment by gamazeps** [comment]
@mboes PTAL
